lib/shrine/plugins/derivation_endpoint.rb

```diff
@@ -0,0 +1,1295 @@
+    #
+    # You can also bump the `:version` per URL:
+    #
+    #     uploaded_file.derivation_url(:thumbnail, version: 1)
+    #     #=> "/thumbnail/eyJpZCI6ImZvbyIsInN?version=1&signature=..."
+    #
+    # ## Accessing uploaded file
+    #
+    # If you want to access the source `UploadedFile` object when deriving, you
+    # can set `:include_uploaded_file` to `true`.
+    #
+    #     plugin :derivation_endpoint, include_uploaded_file: true
+    #
+    # Now the source `UploadedFile` will be passed as the second argument of
+    # the derivation block:
+    #
+    #     derivation :thumbnail do |file, uploaded_file, width, height|
+    #       uploaded_file             #=> #<Shrine::UploadedFile>
+    #       uploaded_file.id          #=> "9a7d1bfdad24a76f9cfaff137fe1b5c7.jpg"
+    #       uploaded_file.storage_key #=> "store"
+    #       uploaded_file.metadata    #=> {}
+    #
+    #       # ...
+    #     end
+    #
+    # By default original metadata that was extracted during attachment isn't
+    # available, to keep the derivation URL as short as possible. However, if
+    # you want to have original metadata available when deriving, you can set
+    # the `:metadata` option to the list of needed metadata values:
+    #
+    #     plugin :derivation_endpoint, metadata: ["filename", "mime_type"]
+    #
+    # Now `filename` and `mime_type` metadata values will be available in the
+    # derivation block:
+    #
+    #     derivation :thumbnail do |file, uploaded_file, width, height|
+    #       uploaded_file.metadata #=>
+    #       # {
+    #       #  "filename" => "nature.jpg",
+    #       #  "mime_type" => "image/jpeg"
+    #       # }
+    #
+    #       uploaded_file.original_filename #=> "nature.jpg"
+    #       uploaded_file.mime_type         #=> "image/jpeg"
+    #
+    #       # ...
+    #     end
+    #
+    # ## Downloading
+    #
+    # When a derivation is requested, the original uploaded file will be
+    # downloaded to disk before the derivation block is called. If you want
+    # to pass in additional storage-specific download options, you can do so
+    # via `:download_options`:
+    #
+    #     plugin :derivation_endpoint, download_options: {
+    #       sse_customer_algorithm: "AES256",
+    #       sse_customer_key:       "secret_key",
+    #       sse_customer_key_md5:   "secret_key_md5",
+    #     }
+    #
+    # When using `Shrine.derivation_endpoint` or `Shrine.derivation_response`,
+    # if the original uploaded file has been deleted, the error the storage
+    # raises when attempting to download it will be propagated by default. You
+    # can choose to have the endpoint convert these errors to 404 responses by
+    # adding them to `:download_errors`:
+    #
+    #     plugin :derivation_endpoint, download_errors: [
+    #       Errno::ENOENT,              # raised by Shrine::Storage::FileSystem
+    #       Aws::S3::Errors::NoSuchKey, # raised by Shrine::Storage::S3
+    #     ]
```

Comment:
What happens if I don't do this i.e. the errors are propagated to?

Reply:
Yeah, the exceptions will be propagated up, so most likely they will be caught by your web server and converted to a 500 error response.